client/my-sites/themes/theme-showcase.jsx

```diff
@@ -357,6 +362,31 @@ class ThemeShowcase extends Component {
 		}
 	};
 
+	recordSearchThemesTracksEvent = ( action, props ) => {
+		let eventName;
+		switch ( action ) {
```

Comment:
I thought a `switch` might be clearer as an allowlist. We can also use an object as mapping 🙂

Reply:
We can simply do this as well:

```
case 'search_clear_icon_click':
case 'search_dropdown_searchby_item_click':
case 'search_dropdown_taxonomy_item_click':
case 'search_dropdown_show_all_button_click':
case 'search_dropdown_show_less_button_click':
	eventName = 'calypso_themeshowcase_' + action;
	break;
}
```

Reply:
But why do we need allowlist in the first place? What's wrong with just calling `recordTracksEvent( 'calypso_themeshowcase_' + eventName, props );`?

Reply:
Following the rule of thumb mentioned in the FG tracks/tracks-for-developers-product-teams/tracks-naming-conventions (Tampermonkey isn't short-handing the URL 🥲)

![Screen Shot 2022-11-24 at 5 51 45 PM](https://user-images.githubusercontent.com/797888/203752886-4c762c5a-51a7-4f51-8ab7-0d58038b4279.png)

Reply:
Ah right but I was more concerned about the "allowlist". So if we were following the FG, should we just spell out the full event name like this?

```
recordTracksEvent( 'calypso_themeshowcase_search_dropdown_searchby_item_click', ...
```

Reply:
Because if we use that allowlist, it defeats the "... hardcoded so they are easily searchable in the codebase", because then every time a dev searches one of these events, the editor will just find the allowlist, not the actual usages :)

Reply:
> Because if we use that allowlist, it defeats the "... hardcoded so they are easily searchable in the codebase", because then every time a dev searches one of these events, the editor will just find the allowlist, not the actual usages :)

I see what you mean, fair point 🙂
On the other hand, it makes sense to me that we are doing tracking in the parent component `<ThemeShowcase/>` since `<SearchThemes/>` and `<KeyedSuggestions/>` are meant to be reusable components so (1) shouldn't have any tracking on their own, and (2) shouldn't have the namespace `calypso_themeshowcase`. 

I think with this approach, even it another dev searches and finds the events, what they see would be a "hint" to where the event is triggered.

Reply:
Ah I see the prolem now!

Hmm.

What if we pass a callback to the child element for each event? Something like:

```
onShowAllButtonClick={() => recordTracksEvent( 'calypso_themeshowcase_search_dropdown_searchby_item_click', ...)}
```

Not sure. @arthur791004 any idea?

Reply:
I think that'd be the most "correct" way of doing this, but unfortunately it would bloat up the code too much for my taste between passing callbacks up and down between parent/child components just for event tracking 🥲

Reply:
Yes, passing a callback function might be the correct way but I agree @taipeicoder's point. A simple approach is to use a universal callback function, e.g: `onElementClick`. However, it's almost the same as the current way. I think we could leave it as it is 🙂

Reply:
Thanks for chiming in @arthur791004! @fushar let me know if the current approach works for you as well, or it's too verbose and would rather have something like you [suggested](https://github.com/Automattic/wp-calypso/pull/70361#discussion_r1031296476).

Reply:
Neither is perfect, so I guess let's use this current approach 😄

Reply:
I am testing the events now